google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,392 @@
+        """
+        return self.resource_rules_map[applies_to]
+
+
+class Rule(object):
+    """Rule properties from the rule definition file. Also finds violations."""
+
+    def __init__(self, rule_name, rule_index, min_retention, max_retention):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule.
+            rule_index (int): The index of the rule.
+            min_retention(int): minimum value of the age in lifecycle.
+            max_retention(int): maximum value of the age in lifecycle.
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.min_retention = min_retention
+        self.max_retention = max_retention
+
+    def generate_bucket_violation(self, bucket):
+        """Generate a violation.
+
+        Args:
+            bucket (Bucket): The bucket that triggers the violation.
+        Returns:
+            RuleViolation: The violation.
+        """
+        data_lifecycle = bucket.get_lifecycle_rule()
+        data_lifecycle_str = json.dumps(data_lifecycle)
+
+        return RuleViolation(
+            resource_name=bucket.name,
+            resource_id=bucket.id,
+            resource_type=bucket.type,
+            full_name=bucket.full_name,
+            rule_name=self.rule_name,
+            rule_index=self.rule_index,
+            violation_type=VIOLATION_TYPE,
+            violation_data=data_lifecycle_str,
+            resource_data=bucket.data,
+        )
+
+    def find_violations(self, res):
+        """Get a generator for violations.
+
+        Args:
+            res (Resource): A class derived from Resource.
+        Returns:
+            Generator: All violations of the resource breaking the rule.
+
+        Raises:
+            ValueError: Raised if the resource type is bucket.
+        """
+
+        if res.type == 'bucket':
+            return self.find_violations_in_bucket(res)
+        raise ValueError(
+            'only bucket is currently supported'
+        )
+
+    def bucket_max_retention_violation(self, bucket):
+        """Get a generator for violations especially for maximum retention
+           It only supports bucket for now, and will work on generalizing
+           in future PRs.
+
+        Args:
+            bucket (bucket): Find violation from the bucket.
+        Yields:
+            RuleViolation: All max violations of the bucket breaking the rule.
+        """
+        if self.max_retention is None:
+            return
+
+        # There should be a condition which guarantees to delete data
+        bucket_lifecycle = bucket.get_lifecycle_rule()
+        if bucket_lifecycle:
+            for lc_item in bucket_lifecycle:
+                if lc_item.get('action', {}).get('type') == 'Delete':
+                    conditions = lc_item.get('condition', {})
+                    age = conditions.get('age')
+                    if age is not None and len(conditions) == 1:
```

Comment:
this can be joined into one statement:
if age is not None and len(conditions) == 1 and age <= self.max_retention